pingcap/tidb-operator
# Refactor start script and support start script v2

### What problem does this PR solve?

Close #4393
### What is changed and how does it work?


* create pkg `pkg/manager/member/startscript/v2`, and mv `template.go` to pkg `pkg/manager/member/startscript/v1`
* separate `render` func to `pkg/manager/member/startscript`
* add `startScriptVersion` in tc spec to choose startscript

### Code changes


### Tests



Deploy new cluster:
* deploy cluster with `startScriptVersion:v1`
* deploy cluster with `startScriptVersion:v1` and tls-enabled
* deploy cluster with `startScriptVersion:v2` 
* deploy cluster with `startScriptVersion:v2 and tls-enabled

For compatibility:
* deploy cluster by v1.3.2 operator, then upgrade it, check that cluster have not rolling upgrade.
* deploy cluster by v1.3.2 operator, then upgrade it and use `startScriptVersion:v1`,  check that the cluster have not rolling upgrade.
* deploy cluster by v1.3.2 operator, then use `startScriptVersion:v2`,  wait for cluster to be ready and check that the args of all components have not changed.

Debug mod:
* Deploy cluster with `startScriptVersion:v2`, and try to enable debug mod for all components.


### Side effects


### Related changes


### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note
Add `startScriptVersion` field to choose startscript for all components 
```
